README.md
## Portal - Full Text Search Web Service

[![crates.io](https://meritbadge.herokuapp.com/portal)](https://crates.io/crates/portal)

### Purpose

The purpose of this service is to be your full text search web service for any JSON client including JavaScript front-ends like React with fetch.

Portal is a full replacement for [ElasticSearch](https://www.elastic.co/) and [MeiliSearch](https://github.com/meilisearch/MeiliSearch). These services are too complex, too resource hungry, and too slow.

Portal is built to be simple and blazing fast with JWT verification, indexing, deindexing, search, and suggest. 

Indexing is batch indexing by default.

To use this service you need to have a running [sonic](https://crates.io/crates/sonic-server) server and a [broker](https://crates.io/crates/broker) server.

### Features

* Very performant with almost no CPU and memory usage
* Supports 87 natural languages
* Under 500 lines of code
* Supports CORS
* Supports JWT authenication
* Multi-tenant
* Supports SSL - full end-to-end encryption
* JSON API
* Auto-provision and renews SSL cert via LetsEncrypt
* Built on [Sonic](https://crates.io/crates/sonic-server) and [Broker](https://crates.io/crates/broker)

### Use

- create a user on [broker](https://crates.io/crates/broker), login, and get a JWT - then attach the JWT as an Authorization: Bearer {token} to the following JSON API endpoints

#### Index

```html
POST /index
```
- authenticated endpoint (Authorization: Bearer {jwt})
example:
```json
{
    "items": [{
        "collection": "coffee", 
        "bucket": "tenant_1", 
        "id": "49e28aae-88d4-4c19-86d8-51f2c9f11039", 
        "data": {
            "name": "roasted",
            "image": "https://img.com/bucket/123/123.jpg"
        },
        "locale": "eng",
        "indexes": ["name"]
    }]
}
```
- note: `locale` is an optional field of an [ISO 639-3 locale code](https://iso639-3.sil.org/code_tables/639/data) - if not defined locale will be auto-detected

- Supported Locales:
    * afr
    * aka
    * amh
    * ara
    * azj
    * bel
    * ben
    * bho
    * bul
    * cat
    * ceb
    * ces
    * cmn
    * dan
    * deu
    * ell
    * eng
    * epo
    * est
    * fin
    * fra
    * guj
    * hat
    * hau
    * heb
    * hin
    * hrv
    * hun
    * ibo
    * ilo
    * ind
    * ita
    * jav
    * jpn
    * kan
    * kat
    * khm
    * kin
    * kor
    * kur
    * lat
    * lav
    * lit
    * mai
    * mal
    * mar
    * mkd
    * mlg
    * mya
    * nep
    * nld
    * nno
    * nob
    * nya
    * ori
    * orm
    * pan
    * pes
    * pol
    * por
    * ron
    * run
    * rus
    * sin
    * skr
    * slk
    * slv
    * sna
    * som
    * spa
    * srp
    * swe
    * tam
    * tel
    * tgl
    * tha
    * tir
    * tuk
    * tur
    * uig
    * ukr
    * urd
    * uzb
    * vie
    * ydd
    * yor
    * zul

will return: `200` or `500` or `400` or `401`

#### Search

```html
POST /search
```
- authenticated endpoint (Authorization: Bearer {jwt})
```json
{
    "collection": "coffee", 
    "bucket": "tenant_1", 
    "query": "roasted",
    "limit": 10,
    "offset": 10
}
```
- note: limit and offset are optional fields

will return: `200` or `500` or `400` or `401`

200 - will return an array of objects
```json
[
    {
        "collection": "coffee", 
        "bucket": "tenant_1", 
        "id": "49e28aae-88d4-4c19-86d8-51f2c9f11039", 
        "data": {
            "name": "roasted",
            "image": "https://img.com/bucket/123/123.jpg"
        },
        "locale": "eng",
        "indexes": ["name"]
    }
]
```

#### Suggest

```html
POST /suggest
```
- authenticated endpoint (Authorization: Bearer {jwt})
```json
{
    "collection": "coffee", 
    "bucket": "tenant_1", 
    "query": "r",
    "limit": 10
}
```
- note: limit is an optional field

will return: `200` or `500` or `400` or `401`

200 - will return an array of words (strings)
```json
["roasted"]
```

#### Deindex

```html
POST /deindex
```
- authenticated endpoint (Authorization: Bearer {jwt})
```json
{
    "ids": ["49e28aae-88d4-4c19-86d8-51f2c9f11039"]
}
```

will return: `200` or `500` or `400` or `401`

### Install

``` cargo install portal ```

- the origin can be passed in as a flag - default `*`
- the port can be passed in as a flag - default `8888` - can only be set for unsecure connections
- the secure flag (https) and can be true or false - default `false`
- the certs flag is the storage path of LetsEncrypt certs - default `certs`
- the db flag is the path where the embedded database will be saved - default `db`
- the domain flag is the domain name (e.g. api.broker.com) of the domain you want to register with LetsEncrypt - must be fully resolvable 
- the sonic_server flag is the sonic domain/ip/port of the sonic server - default `localhost:1491`
- the sonic_password flag is the sonic password set in the sonic config file - default `SecretPassword`
- the broker flag is the broker domain/ip/port of the broker server - default `http://localhost:8080`
- production example: `./portal --secure="true" --domain="index.broker.com" --sonic_server="sonic.broker.com" --sonic_password="wj34T%$Dx" --broker="https://broker.broker.com"`

### Service

There is an example `systemctl` service for Ubuntu called `portal.service` in the code

### TechStack

* [Tide](https://crates.io/crates/tide)
* [RocksDB](https://crates.io/crates/rocksdb)

### Inspiration

* [Broker](https://crates.io/crates/broker)
* [Sonic](https://crates.io/crates/sonic-server)
* [ElasticSearch](https://www.elastic.co/)
* [MeiliSearch](https://github.com/meilisearch/MeiliSearch)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.4] - 2021-04-05

### Updated
- Updated README

## [1.0.3] - 2021-04-05

#### Fixed
- Fixed hanging indexes on re-indexing

### Updated
- Updated README

## [1.0.2] - 2021-04-04

### Updated
- Updated README

## [1.0.1] - 2021-04-04

### Fixed
- Fixed indexing issue with multiple indexes on one item

## [1.0.0] - 2021-04-04

First Release


src/main.rs
use std::{iter::Iterator, str::FromStr};
use serde_derive::{Deserialize, Serialize};
use anyhow::Result;
use surf::http::{Url, Method, Mime};
use lazy_static::lazy_static;
use std::sync::Arc;
use serde_json::json;
use tide::Request;
use http_types::headers::HeaderValue;
use tide::security::{CorsMiddleware, Origin};
use tide_acme::{AcmeConfig, TideRustlsExt};
use sonic_channel::*;

lazy_static! {
    static ref DB : Arc<rocksdb::DB> = {

        let prefix_extractor = rocksdb::SliceTransform::create_fixed_prefix(5);

        let mut opts = rocksdb::Options::default();
        opts.create_if_missing(true);
        opts.set_prefix_extractor(prefix_extractor);

        let configure = env_var_config();
        let db = rocksdb::DB::open(&opts, configure.db).unwrap();
        Arc::new(db)
    };
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Claims {
    pub exp: i64,          
    pub iat: i64,         
    pub iss: String,         
    pub sub: String,
}

#[derive(Deserialize, Debug, Clone)]
pub struct EnvVarConfig {
  pub port: u16,
  pub origin: String,
  pub db: String,
  pub secure: bool,
  pub certs: String,
  pub domain: String,
  pub sonic_server: String,
  pub sonic_password: String,
  pub broker: String,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
pub struct Item {
    pub id: uuid::Uuid,
    pub collection: String,
    pub bucket: String,
    pub data: serde_json::Map<String, serde_json::Value>,
    pub indexes: Vec<String>,
    pub locale: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct IndexForm {
    pub items: Vec<Item>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct DeindexForm {
    pub ids: Vec<uuid::Uuid>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct SearchForm {
    pub query: String,
    pub collection: String,
    pub bucket: String,
    pub limit: Option<usize>,
    pub offset: Option<usize>,
}
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct SuggestForm {
    pub query: String,
    pub collection: String,
    pub bucket: String,
    pub limit: Option<usize>,
}

fn replace(key: String, value: Vec<u8>) -> Result<()> {
    DB.put(key.clone(), value.clone())?;
    Ok(())
}

fn delete(key: String) -> Result<()> {
    DB.delete(key.clone())?;
    Ok(())
}

fn get_items() -> Result<Vec<Item>> {
    let prefix = "items".as_bytes();
    let i = DB.prefix_iterator(prefix);
    let res : Vec<Item> = i.map(|(_, v)| {
        let data: Item = rmp_serde::from_read_ref(&v).unwrap();
        data
    }).collect();
    Ok(res)
}

fn puts_items(items: Vec<Item>) -> Result<()> {
    for item in items {
        let key = format!("items_{}", item.id);
        let value = rmp_serde::to_vec_named(&item)?;
        replace(key, value)?;
    }
    Ok(())
}

fn get_item_by_id(id: uuid::Uuid) -> Result<Option<Item>> {
    let items = get_items()?;
    Ok(items.into_iter().filter(|item| item.id == id).last())
}

fn del_items(ids: Vec<uuid::Uuid>) -> Result<()> {
    for id in ids {
        let key = format!("items_{}", id);
        delete(key)?;
    }
    Ok(())
}

fn env_var_config() -> EnvVarConfig {
 
    let mut port : u16 = 8888;
    let mut secure = false;
    let mut origin = "*".to_string();
    let mut db: String = "db".to_string();
    let mut certs = "certs".to_string();
    let mut domain = "localhost".to_string();
    let mut sonic_server = "localhost:1491".to_string();
    let mut sonic_password = "SecretPassword".to_string();
    let mut broker = "http://localhost:8080".to_string();
    
    let _ : Vec<String> = go_flag::parse(|flags| {
        flags.add_flag("port", &mut port);
        flags.add_flag("origin", &mut origin);
        flags.add_flag("secure", &mut secure);
        flags.add_flag("db", &mut db);
        flags.add_flag("domain", &mut domain);
        flags.add_flag("certs", &mut certs);
        flags.add_flag("sonic_server", &mut sonic_server);
        flags.add_flag("sonic_password", &mut sonic_password);
        flags.add_flag("broker", &mut broker);
    });

    EnvVarConfig{port, origin, secure, domain, certs, db, sonic_server, sonic_password, broker}
}

async fn jwt_verify(token: String) -> Result<bool> {

    let configure = env_var_config();

    let mut parts = token.split(" ");
    let auth_type = parts.next().unwrap();
    if auth_type == "Bearer" {
        let token = parts.next().unwrap();

        let broker_url = format!("{}/verify", configure.broker);
        let auth = format!("Bearer {}", token);
        let url = Url::parse(&broker_url)?;
        let mime = Mime::from_str("application/json").unwrap();
        let request = surf::Request::builder(Method::Get, url.clone())
        .header("authorization", &auth)
        .content_type(mime)
        .build();

        let res = surf::client().send(request).await.unwrap();
        if res.status() == 200 {
            Ok(true)
        } else {
            Ok(false)
        }
    } else {
        Ok(false)
    }
}

fn index_with_sonic(items: Vec<Item>) -> Result<()> {
    let configure = env_var_config();

    let channel = IngestChannel::start(configure.sonic_server, configure.sonic_password)?;
    for item in items {
        match get_item_by_id(item.id)? {
            Some(item) => {
                let mut ids = Vec::new();
                ids.push(item.id);
                deindex_with_sonic(ids)?;
            },
            None => {}
        }
        for (field, value) in item.clone().data {
            for index_field in item.indexes.clone() {
                let oid = format!("{}_{}", item.id.to_string(), index_field);
                if index_field == field && item.clone().locale == None {
                    channel.push(&item.collection, &item.bucket, &oid, &value.to_string())?;
                }
                else if index_field == field && item.clone().locale != None {
                    channel.push_with_locale(&item.collection, &item.bucket, &oid, &value.to_string(), &item.clone().locale.unwrap())?;
                }
            }
        }
    }
    Ok(())
}

fn deindex_with_sonic(ids: Vec<uuid::Uuid>) -> Result<()> {
    let configure = env_var_config();

    let channel = IngestChannel::start(configure.sonic_server, configure.sonic_password)?;
    for id in ids {
        match get_item_by_id(id)? {
            Some(item) => {
                for index in item.indexes {
                    let oid = format!("{}_{}", item.id.to_string(), index);
                    channel.flusho(&item.collection, &item.bucket, &oid)?;
                }
            },
            None => {}
        }
    }
    Ok(())
}

fn search_with_sonic(sf: SearchForm) -> Result<Vec<Item>> {
    let configure = env_var_config();

    let channel = SearchChannel::start(configure.sonic_server, configure.sonic_password)?;

    let mut items = Vec::new();

    if sf.offset != None && sf.limit != None {
        let ids: Vec<String> = channel.query_with_limit_and_offset(&sf.collection, &sf.bucket, &sf.query, sf.limit.unwrap(), sf.offset.unwrap())?;
        for id_str in ids {
            let uid: Vec<&str> = id_str.split("_").collect();
            let id = uuid::Uuid::parse_str(&uid[0])?;
            let item = get_item_by_id(id)?;
            items.push(item.unwrap());
        }
    }
    else if sf.offset == None && sf.limit != None {
        let ids: Vec<String> = channel.query_with_limit(&sf.collection, &sf.bucket, &sf.query, sf.limit.unwrap())?;
        for id_str in ids {
            let uid: Vec<&str> = id_str.split("_").collect();
            let id = uuid::Uuid::parse_str(&uid[0])?;
            match get_item_by_id(id)? {
                Some(item) => {
                    items.push(item);
                },
                None => {}
            }
        }
    }
    else {
        let ids: Vec<String> = channel.query(&sf.collection, &sf.bucket, &sf.query)?;
        for id_str in ids {
            let uid: Vec<&str> = id_str.split("_").collect();
            let id = uuid::Uuid::parse_str(&uid[0])?;
            match get_item_by_id(id)? {
                Some(item) => {
                    items.push(item);
                },
                None => {}
            }
        }
    }

    Ok(items)
}

fn suggest_with_sonic(sf: SuggestForm) -> Result<Vec<String>> {
    let configure = env_var_config();

    let channel = SearchChannel::start(configure.sonic_server, configure.sonic_password)?;

    if sf.limit != None {
        return Ok(channel.suggest_with_limit(&sf.collection, &sf.bucket, &sf.query, sf.limit.unwrap())?);
    }
    else {
        return Ok(channel.suggest(&sf.collection, &sf.bucket, &sf.query)?);
    }
}

async fn index(mut req: Request<()>) -> tide::Result {
    let token_value = req.header("authorization");
    match token_value {
        Some(token_header) => {
            let token = token_header.last().to_string();
            let check = jwt_verify(token).await?;
            if check {
                    let r =  req.body_string().await?;
                    let index_form : IndexForm = serde_json::from_str(&r)?;
                    let items = index_form.items;
                    index_with_sonic(items.clone())?;
                    puts_items(items.clone())?;
                    Ok(tide::Response::builder(200).header("content-type", "application/json").build())
            } else {
                Ok(tide::Response::builder(401).header("content-type", "application/json").build())
            }
        },
        None => { Ok(tide::Response::builder(401).header("content-type", "application/json").build()) }
    }
}

async fn deindex(mut req: Request<()>) -> tide::Result {
    let token_value = req.header("authorization");
    match token_value {
        Some(token_header) => {
            let token = token_header.last().to_string();
            let check = jwt_verify(token).await?;
            if check {
                let r =  req.body_string().await?;
                let deindex_form : DeindexForm = serde_json::from_str(&r)?;
                let ids = deindex_form.ids;
                del_items(ids.clone())?;
                deindex_with_sonic(ids.clone())?;
                Ok(tide::Response::builder(200).header("content-type", "application/json").build())
            } else {
                Ok(tide::Response::builder(401).header("content-type", "application/json").build())
            }
        },
        None => { Ok(tide::Response::builder(401).header("content-type", "application/json").build()) }
    }
}

async fn search(mut req: Request<()>) -> tide::Result {
    let token_value = req.header("authorization");
    match token_value {
        Some(token_header) => {
            let token = token_header.last().to_string();
            let check = jwt_verify(token).await?;
            if check {
                    let r =  req.body_string().await?;
                    let search_form : SearchForm = serde_json::from_str(&r)?;
                    let result = search_with_sonic(search_form)?;
                    Ok(tide::Response::builder(200).body(json!(result)).header("content-type", "application/json").build())
            } else {
                Ok(tide::Response::builder(401).header("content-type", "application/json").build())
            }
        },
        None => { Ok(tide::Response::builder(401).header("content-type", "application/json").build()) }
    }
}

async fn suggest(mut req: Request<()>) -> tide::Result {
    let token_value = req.header("authorization");
    match token_value {
        Some(token_header) => {
            let token = token_header.last().to_string();
            let check = jwt_verify(token).await?;
            if check {
                let r =  req.body_string().await?;
                let suggest_form : SuggestForm = serde_json::from_str(&r)?;
                let result = suggest_with_sonic(suggest_form)?;
                Ok(tide::Response::builder(200).body(json!(result)).header("content-type", "application/json").build())
            } else {
                Ok(tide::Response::builder(401).header("content-type", "application/json").build())
            }
        },
        None => { Ok(tide::Response::builder(401).header("content-type", "application/json").build()) }
    }
}

#[async_std::main]
async fn main() -> tide::Result<()> {

    let configure = env_var_config();

    let cors = CorsMiddleware::new()
    .allow_methods("GET, POST, OPTIONS".parse::<HeaderValue>().unwrap())
    .allow_headers("authorization".parse::<HeaderValue>().unwrap())
    .allow_origin(Origin::from(configure.origin))
    .allow_credentials(false);
    
    let mut app = tide::new();
    app.with(driftwood::DevLogger);
    app.with(cors);
    app.at("/index").post(index);
    app.at("/deindex").post(deindex);
    app.at("/search").post(search);
    app.at("/suggest").post(suggest);

    let ip = format!("0.0.0.0:{}", configure.port);

    if configure.secure {
        app.listen(
            tide_rustls::TlsListener::build().addrs("0.0.0.0:443").acme(
                AcmeConfig::new()
                    .domains(vec![configure.domain])
                    .cache_dir(configure.certs)
                    .production(),
            ),
        )
        .await?;
    } else {
        app.listen(ip).await?;
    }

    Ok(())
}